Userland/Libraries/LibGL/SoftwareGLContext.cpp

```diff
@@ -3317,42 +3317,9 @@ void SoftwareGLContext::build_extension_string()
     m_extensions = String::join(" ", extensions);
 }
 
-void SoftwareGLContext::gl_lightf(GLenum light, GLenum pname, GLfloat param)
+void SoftwareGLContext::gl_light(GLenum light, GLenum pname, GLfloat x, GLfloat y, GLfloat z, GLfloat w)
 {
-    APPEND_TO_CALL_LIST_AND_RETURN_IF_NEEDED(gl_lightf, light, pname, param);
-
-    RETURN_WITH_ERROR_IF(m_in_draw_state, GL_INVALID_OPERATION);
-    RETURN_WITH_ERROR_IF(light < GL_LIGHT0 || light >= (GL_LIGHT0 + m_device_info.num_lights), GL_INVALID_ENUM);
-    RETURN_WITH_ERROR_IF(!(pname == GL_CONSTANT_ATTENUATION || pname == GL_LINEAR_ATTENUATION || pname == GL_QUADRATIC_ATTENUATION || pname != GL_SPOT_EXPONENT || pname != GL_SPOT_CUTOFF), GL_INVALID_ENUM);
-
-    auto& light_state = m_light_states.at(light - GL_LIGHT0);
-
-    switch (pname) {
-    case GL_CONSTANT_ATTENUATION:
-        light_state.constant_attenuation = param;
-        break;
-    case GL_LINEAR_ATTENUATION:
-        light_state.linear_attenuation = param;
-        break;
-    case GL_QUADRATIC_ATTENUATION:
-        light_state.quadratic_attenuation = param;
-        break;
-    case GL_SPOT_EXPONENT:
-        light_state.spotlight_exponent = param;
-        break;
-    case GL_SPOT_CUTOFF:
-        light_state.spotlight_cutoff_angle = param;
-        break;
-    default:
-        VERIFY_NOT_REACHED();
-    }
-
-    m_light_state_is_dirty = true;
-}
-
-void SoftwareGLContext::gl_lightfv(GLenum light, GLenum pname, GLfloat const* params)
-{
-    APPEND_TO_CALL_LIST_AND_RETURN_IF_NEEDED(gl_lightfv, light, pname, params);
+    APPEND_TO_CALL_LIST_AND_RETURN_IF_NEEDED(gl_light, light, pname, x, y, z, w);
     RETURN_WITH_ERROR_IF(m_in_draw_state, GL_INVALID_OPERATION);
     RETURN_WITH_ERROR_IF(light < GL_LIGHT0 || light >= (GL_LIGHT0 + m_device_info.num_lights), GL_INVALID_ENUM);
     RETURN_WITH_ERROR_IF(!(pname == GL_AMBIENT || pname == GL_DIFFUSE || pname == GL_SPECULAR || pname == GL_POSITION || pname == GL_CONSTANT_ATTENUATION || pname == GL_LINEAR_ATTENUATION || pname == GL_QUADRATIC_ATTENUATION || pname == GL_SPOT_CUTOFF || pname == GL_SPOT_EXPONENT || pname == GL_SPOT_DIRECTION), GL_INVALID_ENUM);
```

Comment:
By combining the `gl_light` and `gl_lightfv` calls, we are now accepting more parameters regardless of the way our API was invoked. I.e. `glLight` may only accept "GL_SPOT_EXPONENT, GL_SPOT_CUTOFF, GL_CONSTANT_ATTENUATION, GL_LINEAR_ATTENUATION, and GL_QUADRATIC_ATTENUATION" but we also accept the multi-parameter variants here.

So previously invoking `glLight` with `GL_AMBIENT` would result in an error (correct behavior imho), but now it would overwrite the ambient color.

Reply:
Good point. I think I'll keep these functions as they were and just implement the `i{v}` values for them. Otherwise this is going to be completely unmanageable tangle of janky C++.